Core/include/Acts/Utilities/KDTree.hpp

```diff
@@ -0,0 +1,679 @@
+    //
+    // One interesting thing to note is that all of the nodes in the k-d tree
+    // have a range in the element vector of the outermost node. They simply
+    // make in-place changes to this array, and they hold no memory of their
+    // own.
+    if (m_elems.size() > LeafSize) {
+      m_root = std::make_unique<KDTreeNode>(m_elems.begin(), m_elems.end());
+    } else {
+      m_root = std::make_unique<KDTreeLeaf>(m_elems.begin(), m_elems.end());
+    }
+  }
+
+  /// @brief Perform an orthogonal range search within the k-d tree.
+  ///
+  /// A range search operation is one that takes a k-d tree and an orthogonal
+  /// range, and returns all values associated with coordinates in the k-d tree
+  /// that lie within the orthogonal range. k-d trees can do this operation
+  /// quickly.
+  ///
+  /// @param r The range to search for.
+  ///
+  /// @return The vector of all values that lie within the given range.
+  std::vector<Type> rangeSearch(const range_t &r) const {
+    std::vector<Type> out;
+
+    rangeSearch(r, out);
+
+    return out;
+  }
+
+  /// @brief Perform an orthogonal range search within the k-d tree, returning
+  /// keys as well as values.
+  ///
+  /// Performs the same logic as the keyless version, but includes a copy of
+  /// the key with each element.
+  ///
+  /// @param r The range to search for.
+  ///
+  /// @return The vector of all key-value pairs that lie within the given
+  /// range.
+  std::vector<pair_t> rangeSearchWithKey(const range_t &r) const {
+    std::vector<pair_t> out;
+
+    rangeSearchWithKey(r, out);
+
+    return out;
+  }
+
+  /// @brief Perform an in-place orthogonal range search within the k-d tree.
+  ///
+  /// This range search module operates in place, writing its results to the
+  /// given output vector.
+  ///
+  /// @param r The range to search for.
+  /// @param v The vector to write the output to.
+  void rangeSearch(const range_t &r, std::vector<Type> &v) const {
+    rangeSearchInserter(r, std::back_inserter(v));
+  }
+
+  /// @brief Perform an in-place orthogonal range search within the k-d tree,
+  /// including keys in the result.
+  ///
+  /// Performs the same operation as the keyless version, but includes the keys
+  /// in the results.
+  ///
+  /// @param r The range to search for.
+  /// @param v The vector to write the output to.
+  void rangeSearchWithKey(const range_t &r, std::vector<pair_t> &v) const {
+    rangeSearchInserterWithKey(r, std::back_inserter(v));
+  }
+
+  /// @brief Perform an orthogonal range search within the k-d tree, writing
+  /// the resulting values to an output iterator.
+  ///
+  /// This method allows the user more control in where the result is written
+  /// to.
+  ///
+  /// @tparam OutputIt The type of the output iterator.
+  ///
+  /// @param r The range to search for.
+  /// @param i The iterator to write the output to.
+  template <typename OutputIt>
+  void rangeSearchInserter(const range_t &r, OutputIt i) const {
```

Comment:
for my understanding: does this create a local copy of `i` within the method?

Reply:
Yes. As far as I know passing iterators by value is the canonical way to do that in C++. This sort of makes sense if you think about it, because you have, in essence, four options for passing things to C++ functions:

1. As values, which I've done here.
2. As lvalue references, which would not allow you pass any temporaries to the function, so that would be inconvenient.
3. As const lvalue references, which allow you to pass temporaries but would not allow you to write to the iterator, so that would be useless as well.
4. As rvalue references, which in this case is pretty much the same as passing by value, but with extra steps in this case. Iterators are designed to be super-low overhead, so moving them doesn't really make sense for them.

Reply:
Thanks!